In [120]:
import json
with open('../latest-1000.json') as data_file:
    data=json.load(data_file)

for key in data[0]:
    print(key)
#for i in range(900):
#    print(data[i]['type'])

    
    
print(data[14]['aliases'])

list((map(lambda i: i['id'],data))).index('Q42')

labels
aliases
id
descriptions
claims
sitelinks
type
{'hy': [{'language': 'hy', 'value': 'Ադամս, Դուգլաս'}], 'fr': [{'language': 'fr', 'value': 'Douglas Noel Adams'}, {'language': 'fr', 'value': 'Douglas Noël Adams'}], 'pt': [{'language': 'pt', 'value': 'Douglas Noël Adams'}, {'language': 'pt', 'value': 'Douglas Noel Adams'}], 'ru': [{'language': 'ru', 'value': 'Адамс, Дуглас'}], 'zh': [{'language': 'zh', 'value': '亞當斯'}], 'ja': [{'language': 'ja', 'value': 'ダグラス・アダムス'}], 'nb': [{'language': 'nb', 'value': 'Douglas Noël Adams'}], 'be-tarask': [{'language': 'be-tarask', 'value': 'Дуглас Адамс'}], 'el': [{'language': 'el', 'value': 'Ντάγκλας Νόελ Άνταμς'}], 'it': [{'language': 'it', 'value': 'Douglas Noel Adams'}], 'als': [{'language': 'als', 'value': 'Douglas Noel Adams'}], 'es': [{'language': 'es', 'value': 'Douglas Noel Adams'}, {'language': 'es', 'value': 'Douglas Noël Adams'}], 'pt-br': [{'language': 'pt-br', 'value': 'Douglas Noël Adams'}, {'language': 'pt-br', 'value': 'Douglas No

14

In [121]:

data[14]['claims']['P1005'][0] ['references']

[{'hash': 'a51d6594fee36c7452eaed2db35a4833613a7078',
  'snaks': {'P143': [{'datatype': 'wikibase-item',
     'datavalue': {'type': 'wikibase-entityid',
      'value': {'entity-type': 'item', 'numeric-id': 54919}},
     'property': 'P143',
     'snaktype': 'value'}]},
  'snaks-order': ['P143']}]

In [122]:

def non_wiki_ref(claim):
    wikipedia=['Q52','Q328','Q48183']
    for i in range(len(claim ['references'])):
        if claim ['references'] [i] ['property'] not in wikipedia:
            return True   

In [123]:
def count_references(claims):
    property_list=[]
    
    for p in claims:
        if True:#not wiki_ref(p):
            for i in claims[p]:  
                if 'references' in i:
                    
                    property_list.append(p)
    return property_list


In [124]:
def count_qualifers(claims):
    l=[]
    for p in claims:
        
        
        for i in claims[p]:
            if 'qualifiers' in i:
                l.append(p)
    return l
count_qualifers(data[14]['claims']) 
#set(count_qualifers(data[14]['claims'])).intersection(count_references(data[14]['claims']))

['P998',
 'P998',
 'P998',
 'P998',
 'P998',
 'P998',
 'P1411',
 'P1411',
 'P69',
 'P69',
 'P26',
 'P551',
 'P551']

In [125]:
def alias(data):
    return list(data['aliases'].keys())
     

alias(data[0])  

['ca',
 'fa',
 'te',
 'sv',
 'ml',
 'it',
 'nds',
 'sr',
 'eo',
 'ilo',
 'nl',
 'en',
 'hr',
 'ta',
 'fr',
 'pt',
 'ru',
 'fi',
 'de',
 'lt',
 'es',
 'pl',
 'scn',
 'pt-br',
 'hu',
 'bg',
 'cs',
 'id']

In [126]:
import operator
def score_item(item):
    qualifiers=count_qualifers(item['claims'])
    references=count_references(item['claims'])
    aliases=alias(item)
    return len(qualifiers + references + aliases)

def score_dataset(data):
    result={d['id']:score_item(d) for d in data if d['type']=='item'}
    max_score = max(result[k] for k in result)
    return {k: result[k]/max_score for k in result}

normalized_result = score_dataset(data)

data_humans=[]
for d in range(len(data)):
    HUMAN=5
    for c in list(data[d]['claims'].keys()):
        if (c=='P31'):
            number=data[d]['claims']['P31'][0]['mainsnak']['datavalue']['value']['numeric-id']
            if (number==HUMAN):
                data_humans.append(data[d])
non_humans = [d for d in data if d not in data_humans]   

sorted_ranking = sorted(normalized_result.items(), reverse=True, key=operator.itemgetter(1))
#print(sorted_ranking)

In [131]:
non_human_score = score_dataset(non_humans)
human_score = score_dataset(data_humans)

    
import csv
def ranking_to_csv(filename, ranking):
    with open(filename, "w") as f:
        w = csv.writer(f)
        for item_id in ranking:
            w.writerow([item_id, ranking[item_id]])
ra'non_humans.csv')

In [ ]:
data_humans[11]['id']

{'rank': 'normal', 'references': [{'snaks-order': ['P248', 'P214', 'P813'], 'hash': '050ec907ff2d96e82eddea6ecfc54f12503b9f4c', 'snaks': {'P248': [{'datatype': 'wikibase-item', 'property': 'P248', 'snaktype': 'value', 'datavalue': {'type': 'wikibase-entityid', 'value': {'entity-type': 'item', 'numeric-id': 54919}}}], 'P813': [{'datatype': 'time', 'property': 'P813', 'snaktype': 'value', 'datavalue': {'type': 'time', 'value': {'timezone': 0, 'after': 0, 'time': '+2013-12-07T00:00:00Z', 'calendarmodel': 'http://www.wikidata.org/entity/Q1985727', 'before': 0, 'precision': 11}}}], 'P214': [{'datatype': 'external-id', 'property': 'P214', 'snaktype': 'value', 'datavalue': {'type': 'string', 'value': '113230702'}}]}}, {'snaks-order': ['P248', 'P813', 'P854'], 'hash': '84bf485417f064cf273f05ecbda447612eec7314', 'snaks': {'P248': [{'datatype': 'wikibase-item', 'property': 'P248', 'snaktype': 'value', 'datavalue': {'type': 'wikibase-entityid', 'value': {'entity-type': 'item', 'numeric-id': 20666306}}}], 'P854': [{'datatype': 'url', 'property': 'P854', 'snaktype': 'value', 'datavalue': {'type': 'string', 'value': 'http://data.bnf.fr/ark:/12148/cb11888092r'}}], 'P813': [{'datatype': 'time', 'property': 'P813', 'snaktype': 'value', 'datavalue': {'type': 'time', 'value': {'timezone': 0, 'after': 0, 'time': '+2015-10-10T00:00:00Z', 'calendarmodel': 'http://www.wikidata.org/entity/Q1985727', 'before': 0, 'precision': 11}}}]}}], 'type': 'statement', 'id': 'Q42$F078E5B3-F9A8-480E-B7AC-D97778CBBEF9', 'mainsnak': {'datatype': 'wikibase-item', 'property': 'P31', 'snaktype': 'value', 'datavalue': {'type': 'wikibase-entityid', 'value': {'entity-type': 'item', 'numeric-id': 5}}}}

In [ ]:
#import bz2

#bz_file = bz2.BZ2File("latest-all.json.bz2")
#line_list = bz_file.readlines()

In [ ]:
#len(line_list)

In [ ]:
data[0]['claims'] ['P349'] [0].keys()

In [ ]:
print(data_humans[11]['claims']['P21'])
non_wiki_ref(data_humans[11]['claims']['P21']) 